staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go

```diff
@@ -0,0 +1,224 @@
+		IndefLength:   cbor.IndefLengthForbidden,
+		NilContainers: cbor.NilContainerAsNull,
+		TagsMd:        cbor.TagsAllowed,
+	}.EncMode()
+	if err != nil {
+		panic(err)
+	}
+	return em
+}()
+
+var dm cbor.DecMode = func() cbor.DecMode {
+	dm, err := cbor.DecOptions{
+		DupMapKey:         cbor.DupMapKeyEnforcedAPF,
+		TimeTag:           cbor.DecTagRequired,
+		MaxNestedLevels:   0,
+		MaxArrayElements:  0,
+		MaxMapPairs:       0,
+		IndefLength:       cbor.IndefLengthForbidden,
+		TagsMd:            cbor.TagsAllowed,
+		IntDec:            cbor.IntDecConvertSigned,
+		MapKeyByteString:  cbor.MapKeyByteStringForbidden,
+		ExtraReturnErrors: cbor.ExtraDecErrorUnknownField,
+		DefaultMapType:    reflect.TypeOf(map[string]interface{}(nil)),
+		UTF8:              cbor.UTF8RejectInvalid,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+	return dm
+}()
+
+type metaFactory interface {
+	// Interpret should return the version and kind of the wire-format of the object.
+	Interpret(data []byte) (*schema.GroupVersionKind, error)
+}
+
+type defaultMetaFactory struct{}
+
+var mfdm cbor.DecMode = func() cbor.DecMode {
+	dm, err := cbor.DecOptions{
+		DupMapKey:        cbor.DupMapKeyEnforcedAPF,
+		MaxNestedLevels:  0,
+		MaxArrayElements: 0,
+		MaxMapPairs:      0,
+		IndefLength:      cbor.IndefLengthForbidden,
+		TagsMd:           cbor.TagsAllowed,
+		MapKeyByteString: cbor.MapKeyByteStringForbidden,
+		UTF8:             cbor.UTF8RejectInvalid,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+	return dm
+}()
+
+func (mf *defaultMetaFactory) Interpret(data []byte) (*schema.GroupVersionKind, error) {
+	var tm metav1.TypeMeta
+	if err := mfdm.Unmarshal(data, &tm); err != nil {
+		return nil, fmt.Errorf("unable to determine group/version/kind: %w", err)
+	}
+	actual := tm.GetObjectKind().GroupVersionKind()
+	return &actual, nil
+}
+
+type Serializer struct {
+	metaFactory metaFactory
+	creater     runtime.ObjectCreater
+	typer       runtime.ObjectTyper
+}
+
+func NewSerializer(creater runtime.ObjectCreater, typer runtime.ObjectTyper) *Serializer {
+	return newSerializer(&defaultMetaFactory{}, creater, typer)
+}
+
+func newSerializer(metaFactory metaFactory, creater runtime.ObjectCreater, typer runtime.ObjectTyper) *Serializer {
+	return &Serializer{
+		metaFactory: metaFactory,
+		creater:     creater,
+		typer:       typer,
+	}
+}
+
+func (s *Serializer) Identifier() runtime.Identifier {
+	return "cbor"
+}
+
+func (s *Serializer) Encode(obj runtime.Object, w io.Writer) error {
+	// https://www.rfc-editor.org/rfc/rfc8949.html#name-self-described-cbor
+	if _, err := w.Write([]byte("\xd9\xd9\xf7")); err != nil {
+		return err
+	}
+
+	e := em.NewEncoder(w)
+	if u, ok := obj.(runtime.Unstructured); ok {
+		return e.Encode(u.UnstructuredContent())
+	}
+	return e.Encode(obj)
+}
+
+// gvkWithDefaults returns group kind and version defaulting from provided default
+func gvkWithDefaults(actual, defaultGVK schema.GroupVersionKind) schema.GroupVersionKind {
```

Comment:
reading this, I regret ever allowing gvk to be hinted.  I wonder how much code breaks if we change that on decoding.